### PR TITLE
[WFCORE-4770] Analysis document for the CLI controller boot hook

### PR DESCRIPTION
**Pull request by kabir**
https://issues.redhat.com/browse/WFCORE-4770

### CODE REVIEW

File: controller/WFCORE-4770-boot-cli-script.adoc [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+this in the server log.
+* This is internal functionality to speed up the image start on OpenShift so it is not supported on bare metal (see
+Non-Requirements). If the system properties to trigger this are used, we will always log a WARN message saying this
+is only supported in the OpenShift images, and that any other usage is liable to change in the future.
+
+
+==== Impact in Openshift Builder image
+From a cloud user perspective, this can be summarized in few sentences:
+
+> If the EAP container launch script detects that some configuration is required, it starts the server in admin-mode.
```
Comment (jmesnil) [inline controller/WFCORE-4770-boot-cli-script.adoc]: s/EAP/WildFly

File: controller/WFCORE-4770-boot-cli-script.adoc [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+Once configuration is over, the server is reloaded  for the rare cases) into normal mode. All
+configurations aspects are handled internally by the EAP container. The extensions hook allows users to add their own
```
Comment (jmesnil) [inline controller/WFCORE-4770-boot-cli-script.adoc]: s/EAP/WildFly

File: controller/WFCORE-4770-boot-cli-script.adoc [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+custom CLI operations to the global server configuration.”
+
+It means that we now have a `CONFIG` phase (where the server runs in admin-only mode during the
+boot/configuration), followed by a `RUNNING` phase.
+
+New env vars:
+* `EXECUTE_BOOT_SCRIPT_INVOKER` - to control if the CLI boot time hook is to be used. Set to `true` in the builder image.
```
Comment (jmesnil) [inline controller/WFCORE-4770-boot-cli-script.adoc]: Bullet lists is not displayed properly (missing a new line before the list)

File: controller/WFCORE-4770-boot-cli-script.adoc [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+** If an error occurred during this CLI execution, the server is shutdown.
+** If success, the following logic applies:
+*** If restart is required. `:shutdown(restart=true)` is executed which causes the launch script to restart the server
+with the original arguments. Following the restart of the admin-only server it is reloaded into normal mode.
+*** Otherwise the server is reloaded into normal mode.
+* The CLI process terminates.
+
+==== Impact on readiness probe and Operator
+
+While running in admin-mode the server shouldn’t be advertised as ready.
```
Comment (jmesnil) [inline controller/WFCORE-4770-boot-cli-script.adoc]: This feature has an impact on our existing [EAP probe](https://github.com/jboss-container-images/jboss-eap-modules/blob/66f3a7d08876a8e90a337bde5fb2abf3b29bf78c/os-eap-probes/common/added/probes/probe/eap/dmr.py#L38).
The probe UP if the `server-state` is `running` but does not look at the value of the `running-mode` attribute (which can be either `NORMAL` or `ADMIN_ONLY`). 

The probe would have to be updated when this feature is activated.

File: controller/WFCORE-4770-boot-cli-script.adoc (hunk 3 above)
Comment (kabir) [reply controller/WFCORE-4770-boot-cli-script.adoc]: Fixed

File: controller/WFCORE-4770-boot-cli-script.adoc (hunk 4 above)
Comment (kabir) [reply controller/WFCORE-4770-boot-cli-script.adoc]: @jmesnil I added ^^ to the proposal

**Pull request merged**